File: Android_Zone_Lib/src/com/zone/lib/utils/data/file2io2data/ZSharePerference.java
package com.zone.lib.utils.data.file2io2data;

import android.content.Context;
import android.content.SharedPreferences;

import java.lang.reflect.InvocationTargetException;
import java.lang.reflect.Method;
import java.util.Map;
import com.zone.lib.Configure;
import com.zone.lib.utils.data.check.StringCheck;
import com.zone.lib.utils.data.convert.GsonUtils;

/**
 * 引用项目:https://github.com/openproject/LessCode
 * 不习惯$符号的方法名,所以改了下；
 */
public class ZSharePerference {

    private String shareName = "preferences";

    protected ZSharePerference(String shareName) {
        this.shareName = shareName;
    }

    private SharedPreferences getSharedPreferences() {
        return Configure.getApplicationContext()
                .getSharedPreferences(shareName, Context.MODE_PRIVATE);
    }

    public <T> void put(String key, T value) {
        SharedPreferences sp = getSharedPreferences();
        SharedPreferences.Editor editor = sp.edit();

        if (value == null) {
            // if value is null, just handler it as a String
            editor.putString(key, null);
        } else {
            if (value.getClass() == Boolean.class) {
                editor.putBoolean(key, (Boolean) value);
            } else if (value.getClass() == Float.class) {
                editor.putFloat(key, (Float) value);
            } else if (value.getClass() == Integer.class) {
                editor.putInt(key, (Integer) value);
            } else if (value.getClass() == Long.class) {
                editor.putLong(key, (Long) value);
            } else if (value.getClass() == String.class) {
                editor.putString(key, (String) value);
            } else {
                editor.putString(key, GsonUtils.toJson(value));
//                throw new RuntimeException("the put value type can't support.");
            }
        }
        SharedPreferencesCompat.apply(editor);
    }

    public <T> T get(String key, Class<T> t) {
        SharedPreferences sp = getSharedPreferences();
        String content = sp.getString(key, "");
        if (StringCheck.isEmptyTrim(content))
            return null;
        else
            return GsonUtils.fromJson(content, t);
    }

    public String get(String key, String defaultValue) {
        SharedPreferences sp = getSharedPreferences();
        return sp.getString(key, defaultValue);
    }

    public boolean get(String key, boolean defaultValue) {
        SharedPreferences sp = getSharedPreferences();
        return sp.getBoolean(key, defaultValue);
    }

    public float get(String key, float defaultValue) {
        SharedPreferences sp = getSharedPreferences();
        return sp.getFloat(key, defaultValue);
    }

    public int get(String key, int defaultValue) {
        SharedPreferences sp = getSharedPreferences();
        return sp.getInt(key, defaultValue);
    }

    public long get(String key, long defaultValue) {
        SharedPreferences sp = getSharedPreferences();
        return sp.getLong(key, defaultValue);
    }


    public Map<String, ?> getAll() {
        return getSharedPreferences().getAll();
    }

    public void registerOnSharedPreferenceChangeListener(SharedPreferences.OnSharedPreferenceChangeListener listener) {
        getSharedPreferences().registerOnSharedPreferenceChangeListener(listener);
    }

    public void unregisterOnSharedPreferenceChangeListener(SharedPreferences.OnSharedPreferenceChangeListener listener) {
        getSharedPreferences().unregisterOnSharedPreferenceChangeListener(listener);
    }

    public boolean contains(String key) {
        return getSharedPreferences().contains(key);
    }

    public void remove(String key) {
        SharedPreferences sp = getSharedPreferences();
        SharedPreferences.Editor editor = sp.edit();
        editor.remove(key);
        SharedPreferencesCompat.apply(editor);
    }

    public void clear() {
        SharedPreferences sp = getSharedPreferences();
        SharedPreferences.Editor editor = sp.edit();
        editor.clear();
        SharedPreferencesCompat.apply(editor);
    }


    /**
     * *********************************************************************************************
     * Unlike commit(), which writes its preferences out to persistent storage synchronously,
     * apply() commits its changes to the in-memory SharedPreferencesimmediately
     * but starts an asynchronous commit to disk and you won't be notified of any failures.
     * If another editor on this SharedPreferences does a regularcommit() while a apply() is still outstanding,
     * the commit() will block until all async commits are completed as well as the commit itself.
     * *********************************************************************************************
     */
    private static class SharedPreferencesCompat {

        private static final Method sApplyMethod = findApplyMethod();

        /**
         * check apply mthod by reflect
         *
         * @return
         */
        @SuppressWarnings({"unchecked", "rawtypes"})
        private static Method findApplyMethod() {
            try {
                Class clz = SharedPreferences.Editor.class;
                return clz.getMethod("apply");
            } catch (NoSuchMethodException e) {
            }

            return null;
        }

        /**
         * if it has apply(), use apply() first;
         * else just use the commit().
         *
         * @param editor
         */
        public static void apply(SharedPreferences.Editor editor) {
            try {
                if (sApplyMethod != null) {
                    sApplyMethod.invoke(editor);
                    return;
                }
            } catch (IllegalArgumentException | IllegalAccessException | InvocationTargetException e) {
                e.printStackTrace();
            }
            editor.commit();
        }
    }
}

File: Android_Zone_Lib/src/com/zone/lib/utils/data/file2io2data/ZSharePerference.kt
package com.zone.lib.utils.data.file2io2data

import android.content.Context
import android.content.SharedPreferences

import java.lang.reflect.InvocationTargetException
import java.lang.reflect.Method
import com.zone.lib.Configure
import com.zone.lib.LogZSDK
import com.zone.lib.utils.data.check.StringCheck
import com.zone.lib.utils.data.convert.GsonUtils
import java.util.*

/**
 * 引用项目:https://github.com/openproject/LessCode
 * 不习惯$符号的方法名,所以改了下；
 */
open class ZSharePerference protected constructor(val shareName: String) {

    private val weakHashMap: WeakHashMap<Any, Any> by lazy {
        WeakHashMap<Any, Any>()
    }

    private val sharedPreferences: SharedPreferences
        get() = Configure.getApplicationContext()
                .getSharedPreferences(shareName, Context.MODE_PRIVATE)

    val all: Map<String, *>
        get() = sharedPreferences.all


    fun <T> put(key: String, value: T?) {
        val sp = sharedPreferences
        val editor = sp.edit()

        weakHashMap.put(key, value)

        if (value == null) {
            // if value is null, just handler it as a String
            editor.putString(key, null)
        } else {
            if (value is Boolean) {
                editor.putBoolean(key, (value as Boolean?)!!)
            } else if (value is Float) {
                editor.putFloat(key, (value as Float?)!!)
            } else if (value is Int) {
                editor.putInt(key, (value as Int?)!!)
            } else if (value is Long) {
                editor.putLong(key, (value as Long?)!!)
            } else if (value is String) {
                editor.putString(key, value as String?)
            } else {
                editor.putString(key, GsonUtils.toJson(value))
                //throw new RuntimeException("the put value type can't support.");
            }
        }
        SharedPreferencesCompat.apply(editor)
    }

    operator fun <T> get(key: String, t: Class<T>): T? {
        val value = getValueFromCache<T>(key)
        return value ?: getValueByClassInner(key, t)
    }

    private fun <T> getValueFromCache(key: String): T? {
        val value = (weakHashMap.get(key) as? T).apply {
            LogZSDK.d("from cache: key:$key \t value:$this")
        }
        return value
    }

    private fun <T> getValueByClassInner(key: String, t: Class<T>): T? {
        val content = sharedPreferences.getString(key, "")
        return if (StringCheck.isEmptyTrim(content)) null
        else GsonUtils.fromJson(content, t)
    }

    operator fun <T> get(key: String, defaultValue: T): T {
        val value = getValueFromCache<T>(key)
        return value ?: getValueInner(defaultValue, key)
    }

    private fun <T> getValueInner(defaultValue: T, key: String): T {
        var result: T = defaultValue
        when (defaultValue) {
            is String -> {
                result = sharedPreferences.getString(key, defaultValue) as T
            }
            is Boolean -> {
                result = sharedPreferences.getBoolean(key, defaultValue) as T
            }
            is Float -> {
                result = sharedPreferences.getFloat(key, defaultValue) as T
            }
            is Int -> {
                result = sharedPreferences.getInt(key, defaultValue) as T
            }
            is Long -> {
                result = sharedPreferences.getLong(key, defaultValue) as T
            }
            else -> {
                throw  RuntimeException("the put value type can't support.");
            }
        }
        return result
    }

    fun registerOnSharedPreferenceChangeListener(listener: SharedPreferences.OnSharedPreferenceChangeListener) {
        sharedPreferences.registerOnSharedPreferenceChangeListener(listener)
    }

    fun unregisterOnSharedPreferenceChangeListener(listener: SharedPreferences.OnSharedPreferenceChangeListener) {
        sharedPreferences.unregisterOnSharedPreferenceChangeListener(listener)
    }

    operator fun contains(key: String): Boolean {
        return sharedPreferences.contains(key)
    }

    fun remove(key: String) {
        val editor = sharedPreferences.edit()
        editor.remove(key)
        SharedPreferencesCompat.apply(editor)
    }

    fun clear() {
        val editor = sharedPreferences.edit()
        editor.clear()
        SharedPreferencesCompat.apply(editor)
    }


    /**
     * *********************************************************************************************
     * Unlike commit(), which writes its preferences out to persistent storage synchronously,
     * apply() commits its changes to the in-memory SharedPreferencesimmediately
     * but starts an asynchronous commit to disk and you won't be notified of any failures.
     * If another editor on this SharedPreferences does a regularcommit() while a apply() is still outstanding,
     * the commit() will block until all async commits are completed as well as the commit itself.
     * *********************************************************************************************
     */
    private object SharedPreferencesCompat {

        private val sApplyMethod = findApplyMethod()

        /**
         * check apply mthod by reflect
         *
         * @return
         */
        private fun findApplyMethod(): Method? {
            try {
                val clz = SharedPreferences.Editor::class.java
                return clz.getMethod("apply")
            } catch (e: NoSuchMethodException) {
            }

            return null
        }

        /**
         * if it has apply(), use apply() first;
         * else just use the commit().
         *
         * @param editor
         */
        fun apply(editor: SharedPreferences.Editor) {
            try {
                if (sApplyMethod != null) {
                    sApplyMethod.invoke(editor)
                    return
                }
            } catch (e: IllegalArgumentException) {
                e.printStackTrace()
            } catch (e: IllegalAccessException) {
                e.printStackTrace()
            } catch (e: InvocationTargetException) {
                e.printStackTrace()
            }

            editor.commit()
        }
    }
}
